### PR TITLE
DEIMOS Optical Model

### PR DESCRIPTION
**Pull request by kbwestfall**
This may have been more trouble than it's worth, but I've finished transcribing the IDL code with the DEIMOS optical model.  This was the first step in checking the slit locations found by pypeit against their expected positions from the slit mask.  I'll be working on the details of that checking toward the end of next week.  This PR doesn't actually touch anything executed in PypeIt yet, it just introduces the abstract classes and the specific classes used for DEIMOS.  Changes are largely confined to files in the `pypeit/spectrographs` directory.  I should spend some more time documenting the classes/functions, but the code is ready to go.

Changes include:

 - Implementation of `SlitMask` class
 - Implementation of `DetectorMap` class
 - Implementation of `OpticalModel` class
 - Changes to `KeckDEIMOSSpectrograph` class to use the above
 - Includes unit tests for all of the above
 - Adds the notebook [here](https://github.com/pypeit/PypeIt/blob/match_mask/doc/nb/DEIMOSOpticalModel.ipynb) that demos how these can be used

Tests pass.

### CODE REVIEW

File: pypeit/spectrographs/keck_deimos.py [hunk 1]
```diff
@@ -426,6 +439,255 @@ def setup_arcparam(self, arcparam, disperser=None, fitstbl=None, arc_idx=None,
...
+                            hdu['BluSlits'].data['slitX3'],
+                            hdu['BluSlits'].data['slitY3'],
+                            hdu['BluSlits'].data['slitX4'],
+                            hdu['BluSlits'].data['slitY4']]).T.reshape(-1,4,2)
+        self.slitmask = SlitMask(corners, slitid=hdu['BluSlits'].data['dSlitId'])
+        return self.slitmask
+
+    def get_grating(self, filename):
+        """
+        Take from xidl/DEEP2/spec2d/pro/deimos_omodel.pro and
```
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py]: Taken

File: pypeit/spectrographs/keck_deimos.py [hunk 2]
```diff
@@ -426,6 +439,255 @@ def setup_arcparam(self, arcparam, disperser=None, fitstbl=None, arc_idx=None,
...
+        slider = hdu[0].header['GRATEPOS']
+        # TODO: Add test for slider
+
+        # Central wavelength, grating angle, and tilt position
+        if slider == 3:
+            central_wave = hdu[0].header['G3TLTWAV']
+            # Not used
+            #angle = (hdu[0].header['G3TLTRAW'] + 29094)/2500
+            tilt = hdu[0].header['G3TLTVAL']
+        else:
```
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py]: should we make this 

``` 
elif slider in [2,4]:
```

and than have an `else` catching any unexpected value?

File: pypeit/spectrographs/keck_deimos.py [hunk 3]
```diff
@@ -426,6 +439,255 @@ def setup_arcparam(self, arcparam, disperser=None, fitstbl=None, arc_idx=None,
...
+                                     1200: [-0.052,  0.122, 6.9e-4, -0.294],
+                                  'other': [-0.050,  0.080, 6.9e-4, -0.250] } }
+
+        # Return calbirated roll, yaw, and tilt
+        return orientation_coeffs[slider][_ruling][0], \
+                orientation_coeffs[slider][_ruling][1], \
+                tilt*(1-orientation_coeffs[slider][_ruling][2]) \
+                    + orientation_coeffs[slider][_ruling][3]
+
+    def mask_to_pixel_coordinates(self, x=None, y=None, wave=None, order=1, filename=None,
```
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py]: add a doc string

File: pypeit/spectrographs/opticalmodel.py [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+
+        # Return vectors transformed out of the grating conjugate surface
+        return OpticalModel.conjugate_surface_transform(_r, self.transform)
+
+
+# ----------------------------------------------------------------------
+# Vanilla imaging spectrograph optical model
+class OpticalModel:
+    """
+    Vanilla optical model for a imaging spectrograph.
```
Comment (profxj) [inline pypeit/spectrographs/opticalmodel.py]: an

File: pypeit/spectrographs/keck_deimos.py (hunk 1 above)
Comment (kbwestfall) [reply pypeit/spectrographs/keck_deimos.py]: Done.

File: pypeit/spectrographs/keck_deimos.py (hunk 2 above)
Comment (kbwestfall) [reply pypeit/spectrographs/keck_deimos.py]: Done.  Does anyone know if there are more than these three positions?

File: pypeit/spectrographs/keck_deimos.py (hunk 3 above)
Comment (kbwestfall) [reply pypeit/spectrographs/keck_deimos.py]: Done.

File: pypeit/spectrographs/opticalmodel.py (hunk 4 above)
Comment (kbwestfall) [reply pypeit/spectrographs/opticalmodel.py]: Done.

**Comment by kbwestfall** [comment]
Running tests now.  I'll merge, assuming they pass.

**Comment by bpholden** [comment]
Hi Kyle,

You specifically asked if the DEIMOS slider had "3 positions".

The slider has 4 slots, but only 3 and 4 can have a grating. Either 1 or 2
has a mirror (and only a mirror can live in that spot) while the remaining
spot does not work and cannot be occupied.

Brad

**Pull request merged**